src/hotspot/share/gc/shenandoah/shenandoahSimpleBitMap.cpp

```diff
@@ -0,0 +1,244 @@
+           start_idx, count);
+    assert(start_idx + count <= (idx_t) _num_bits, "precondition");
+    size_t array_idx = start_idx >> LogBitsPerWord;
+    uintx bit_number = start_idx & right_n_bits(LogBitsPerWord);
+    uintx element_bits = _bitmap[array_idx];
+    uintx bits_to_examine  = BitsPerWord - bit_number;
+    element_bits >>= bit_number;
+    uintx complement = ~element_bits;
+    uintx trailing_ones;
+    if (complement != 0) {
+      trailing_ones = count_trailing_zeros<uintx>(complement);
+    } else {
+      trailing_ones = bits_to_examine;
+    }
+    if (trailing_ones >= (uintx) count) {
+      return true;
+    } else if (trailing_ones == bits_to_examine) {
+      start_idx += bits_to_examine;
+      count -= bits_to_examine;
+      // Repeat search with smaller goal
+    } else {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool ShenandoahSimpleBitMap::is_backward_consecutive_ones(idx_t last_idx, idx_t count) const {
+  while (count > 0) {
+    assert((last_idx >= 0) && (last_idx < _num_bits), "precondition");
+    assert(last_idx - count >= -1, "precondition");
+    size_t array_idx = last_idx >> LogBitsPerWord;
+    uintx bit_number = last_idx & right_n_bits(LogBitsPerWord);
+    uintx element_bits = _bitmap[array_idx];
+    uintx bits_to_examine = bit_number + 1;
+    element_bits <<= (BitsPerWord - bits_to_examine);
+    uintx complement = ~element_bits;
+    uintx leading_ones;
+    if (complement != 0) {
+      leading_ones = count_leading_zeros<uintx>(complement);
+    } else {
+      leading_ones = bits_to_examine;
+    }
+    if (leading_ones >= (uintx) count) {
+      return true;
+    } else if (leading_ones == bits_to_examine) {
+      last_idx -= leading_ones;
+      count -= leading_ones;
+      // Repeat search with smaller goal
+    } else {
+      return false;
+    }
+  }
+  return true;
+}
+
+idx_t ShenandoahSimpleBitMap::find_first_consecutive_set_bits(idx_t beg, idx_t end, size_t num_bits) const {
+  assert((beg >= 0) && (beg < _num_bits), "precondition");
+
+  // Stop looking if there are not num_bits remaining in probe space.
+  idx_t start_boundary = end - num_bits;
+  if (beg > start_boundary) {
+    return end;
+  }
+  uintx array_idx = beg >> LogBitsPerWord;
+  uintx bit_number = beg & right_n_bits(LogBitsPerWord);
+  uintx element_bits = _bitmap[array_idx];
+  if (bit_number > 0) {
+    uintx mask_out = right_n_bits(bit_number);
+    element_bits &= ~mask_out;
+  }
+  while (true) {
+    if (element_bits == 0) {
+      // move to the next element
+      beg += BitsPerWord - bit_number;
+      if (beg > start_boundary) {
+        // No match found.
+        return end;
+      }
+      array_idx++;
+      bit_number = 0;
+      element_bits = _bitmap[array_idx];
+    } else if (is_forward_consecutive_ones(beg, num_bits)) {
+      return beg;
+    } else {
+      // There is at least one non-zero bit within the masked element_bits.  Find it.
```

Comment:
I think you meant to say 'There is at least one zero bit...' ?

Reply:
The if clause handles the case when element_bits == 0.  The fact that we didn't take the if branch tells us there is at least one non-zero bit.  The else-if clause handles the case that the desired number of consecutive one bits are found at offset beg.  At this point, we know there is at least one non-zero bit and we know that we don't have num_bits consecutive one bits starting at beg.

I'm adding a comment to the code to explain how this works.